File: rician_K.m
function K = rician_K(E_deg)
%RICIAN_K  Elevation-dependent Rician K-factor for LEO satellite links.
%
%   K = rician_K(E_deg) returns the Rician K-factor given an elevation angle.
%
%   The model is fitted cubic splines for Kim et al. (2006)'s measured K values at {20, 30, 40, 60, 80} degrees elevation.
%   The paper's polynomial fit formula contained a rounding error and does not extrapolate well.
%    - 20 < E < 80: cubic spline interpolation
%    - E > 80:      cubic spline extrapolation (further increasing)
%    - E < 20:      fixed to K(20) = 3.07 as extrapolation would result in increasing K
%   The interpolating function can be visualized with scripts/rician_K_plot.m

arguments
    E_deg   % Elevation angle (degrees)
end

persistent K_model
if isempty(K_model)
    E_data = [20, 30, 40, 60, 80];
    K_data = [3.07, 3.24, 3.60, 5.63, 17.06];
    K_model = griddedInterpolant(E_data, K_data, 'spline', 'spline');
end

% Floor below 20 deg: clamp input so the spline returns K(20) for any E < 20 deg
K = K_model(max(E_deg, 20));
end

File: python/lora_dts/rician_k.py
"""Elevation-dependent Rician K-factor, port of rician_K.m."""

import numpy as np
from scipy.interpolate import CubicSpline

_E_DATA = [20.0, 30.0, 40.0, 60.0, 80.0]
_K_DATA = [3.07, 3.24, 3.60, 5.63, 17.06]

# Not-a-knot cubic spline with cubic extrapolation: the same scheme as
# MATLAB griddedInterpolant(E_data, K_data, 'spline', 'spline').
_K_MODEL = CubicSpline(_E_DATA, _K_DATA, bc_type="not-a-knot",
                       extrapolate=True)


def rician_k(e_deg):
    """Rician K-factor for LEO satellite links at elevation e_deg (degrees).

    Fitted cubic spline to Kim et al. (2006)'s measured K values at
    {20, 30, 40, 60, 80} degrees elevation:
      - 20 < E < 80: cubic spline interpolation
      - E > 80:      cubic spline extrapolation (further increasing)
      - E < 20:      fixed to K(20) = 3.07, as extrapolation would result
                     in increasing K
    """
    e = np.maximum(np.asarray(e_deg, dtype=float), 20.0)
    k = _K_MODEL(e)
    return k.item() if np.ndim(e_deg) == 0 else k
